### ISSUE TITLE
Strange regression involving Spice models in the MSL

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#13010]
I am checking the regressions between [Modelica 4.0.0](https://libraries.openmodelica.org/branches/master/Modelica_4.0.0/Modelica_4.0.0.html) and [Modelica 4.1.0](https://libraries.openmodelica.org/branches/master/Modelica_4.1.0-beta.om/Modelica_4.1.0-beta.om.html).

At the moment there are 7 regressions, of which 6 are in the Spice library:
```
Modelica.Electrical.Spice3.Examples.Inverter
Modelica.Electrical.Spice3.Examples.FourInverters
Modelica.Electrical.Spice3.Examples.InvertersApartRecord
Modelica.Electrical.Spice3.Examples.InvertersExtendedModel
Modelica.Electrical.Spice3.Examples.Nor
Modelica.Electrical.Spice3.Examples.Spice3BenchmarkFourBitBinaryAdder
```
At the moment I just analyzed the `Inverter` model, as it is quite likely that the root cause is the same for all. The MSL 4.0.0 model [runs successfully](https://libraries.openmodelica.org/branches/master/Modelica_4.0.0/files/Modelica_4.0.0_Modelica.Electrical.Spice3.Examples.Inverter.sim), while the 4.1.0 [fails](https://libraries.openmodelica.org/branches/master/Modelica_4.1.0-beta.om/files/Modelica_4.1.0-beta.om_Modelica.Electrical.Spice3.Examples.Inverter.sim) with

```
LOG_SUCCESS       | info    | The initialization finished successfully without homotopy method.
LOG_ASSERT        | debug   | division by zero at time 1.025512756378189e-13, (a=inf) / (b=0), where divisor b expression is: v.Tfalling - v.Twidth
```
As far as I can see, the source code `Inverter` and of all its dependencies _has not changed_ between MSL 4.0.0 and 4.1.0. However, if I flatten it in both cases and run diff on the flattened model, there is one difference: in 4.0.0 all calls to `Modelica.Math.log()` are turned into calls to the built-in `log()` function, while in 4.1.0 they are kept as `Modelica.Math.log()`.

Q1: why does this happen in the first place?
Q2: why should it matter for code generation?
Q3: why does it matter so much as to cause the simulation to fail immediatly after initialization?

At the moment I haven't got deeper into the analyis. @perost could you shed some light about Q1? Then, I can try to dig deeper into the remaining two questions.

**Comment by casella** [comment]
The plot thickens: the result actually depends on the operating system (?!?)

On Windows 11, I get this outcome:
```
The initialization finished successfully without homotopy method.
Homotopy solver Newton iteration: Maximum number of iterations reached at time 0.000000, but no root found.
Homotopy solver Newton iteration: Maximum number of iterations reached at time 0.000000, but no root found.
### STATISTICS ###
The simulation finished successfully.
```
On Linux, I get
```
Simulation process failed. Exited with code 255.
division by zero at time 1.050525262631316e-13, (a=inf) / (b=0), where divisor b expression is: v.Tfalling - v.Twidth
```
On Windows, the two parameters get values `v.Tfalling = 1.2e-12 `and `v.Twidth = 1.1e-12`.

I checked with the declarative debugger, where I find
```
111 (assign) v.TF := v.TR
114 (assign) v.Tfalling := v.Twidth + v.TF
```
Unfortunately I can't find any equation for v.TR in the debugger, but I see this code
```
Sources.V_pulse v(V2=5, TR=0.1e-12)
```
so eventually `v.Tfalling - v.Twidth` should be equal to 0.1e-12. Not entirely sure why it isn't on Linux and why this should only be relevant for MSL 4.1.0.

**Comment by casella** [comment]
Here is an MWE:
```modelica
model Test
  Modelica.Electrical.Spice3.Basic.Ground ground;
  Modelica.Electrical.Spice3.Sources.V_pulse v(V2 = 5, TR = 0.1e-12);
equation
  connect(v.n, ground.p);
  annotation(
    experiment(StopTime = 1e-11, Interval = 5e-15, Tolerance = 1e-7),
    uses(Modelica(version = "4.1.0")));
end Test;
```
works on Windows, fails on Linux with
```
division by zero at time 1.050525262631316e-13, (a=inf) / (b=0), where divisor b expression is: v.Tfalling - v.Twidth
```

**Comment by casella** [comment]
OK, maybe I've got a clue on the MWE. Logarithms are actually not involved at all. This is an excerpt of the source code of the Modelica.Electrical.Spice3.Sources.V_pulse model:
```modelica
...
  parameter SI.Time TD = 0.0 "Delay time";
  parameter SI.Time TR(start=1) "Rise time";
  parameter SI.Time TF = TR "Fall time";
  parameter SI.Time PW = Modelica.Constants.inf "Pulse width";
  parameter SI.Time PER= Modelica.Constants.inf "Period";
...
protected
  parameter SI.Time Trising=TR "End time of rising phase within one period";
  parameter SI.Time Twidth=Trising + PW
    "End time of width phase within one period";
  parameter SI.Time Tfalling=Twidth + TF
    "End time of falling phase within one period";
...
equation
...
  v = V1 + (if (time < TD or counter2 == 0 or time >= T0 +
    Tfalling) then 0 else if (time < T0 + Trising) then (time - T0)*
    (V2-V1)/Trising else if (time < T0 + Twidth) then V2-V1 else
    (T0 + Tfalling - time)*(V2-V1)/(Tfalling - Twidth));
```
Now, when the Inverter model is instantiated, v.TR is set to 0.1e-12, but `v.PW` and `v.PER` are left unchanged to their Inf value. This is a bit fishy, because `Twidth` and `Tfalling` are both equal to `Modelica.Constants.Inf`, and then they are subtracted in `(Tfalling - Twidth)`. Not going to end well if that divisor is actually ever called in the conditional expression.

Now, on Windows `Modelica.Constants.Inf` is evaluated to 1e60 both in MSL 4.0.0 and in MSL 4.1.0, and for some reason it does work, despite the fishiness.

On Linux, `Modelica.Constants.Inf` is evaluated to 1e60 in 4.0.0 and to 1.7976931348623157e308 in 4.1.0, which breaks the whole thing.

Now, in Modelica.Constants.Inf is defined as ModelicaServices.Machine.inf, and ModelicaServices.Machine.inf is defined as 1e60 in 4.0.0 and as 1.7976931348623157e308 in 4.1.0.

Why did we change Inf from 1e60 to 1.7e308 in 4.1.0, but only in Linux?

**Comment by casella** [comment]
In any case, the usage of the V_pulse model is fishy, I'll discuss that within MAP-Lib

**Comment by bilderbuchi** [comment]
> On Linux, `Modelica.Constants.Inf` is evaluated to 1e60 in 4.0.0 and to 1.7976931348623157e308 in 4.1.0, which breaks the whole thing.
> 
> Now, in Modelica.Constants.Inf is defined as ModelicaServices.Machine.inf, and ModelicaServices.Machine.inf is defined as 1e60 in 4.0.0 and as 1.7976931348623157e308 in 4.1.0.
> 
> Why did we change Inf from 1e60 to 1.7e308 in 4.1.0, but only in Linux?

1.7976931348623157e308 sounds a lot like the maximum representable floating point value in 64 bit floats.
Infinity in IEE-754 is represented differently, though, I don't know why double_max would be used instead -- this probably messes up NAN propagation etc?

**Comment by casella** [comment]
Huge discussion, see modelica/ModelicaStandardLibrary#4478 and modelica/ModelicaStandardLibrary#4479.

My suggested action at the moment: leave Modelica.Constants.inf alone and just fix the defaults for the Spice3 models with something appropriate, see PR modelica/ModelicaStandardLibrary#4481

**Comment by casella** [comment]
After lengthy discussions in the MA context, I came to the conclusion that the best course of action for the moment being is that we change the OMC-specific implementation of ModelicaServices 4.1.0 to contain
```modelica
  final constant Real small=1e-60
    "Smallest number such that small and -small are representable on the machine";
  final constant Real inf=1e60
    "Biggest Real number such that inf and -inf are representable on the machine";
```
despite instead of the default implementation in [MSL 4.1.0](https://github.com/modelica/ModelicaStandardLibrary/blob/7bec602cba25c2e12ca656519be0e28eaa7afca0/ModelicaServices/package.mo#L179) being
```modelica
    final constant Real small=2.2250738585072014e-308
      "Minimum normalized positive floating-point number";
    final constant Real inf=1.7976931348623157E+308
      "Maximum representable finite floating-point number";
```
This basically keeps the old behaviour of MSL 4.0.0, which was just fine, and resolve this issue.

We could try to figure out why the new inf value causes problems to the rutime, but I understand this has much lower priority than a lot of other more important issues that we need to resolve first. 1e60 is large enough for all practical purposes. No need to wast time on that.

@adrpo can you please take care of that? Thanks!

**Comment by casella** [comment]
Should be fixed by https://github.com/OpenModelica/OpenModelica-ModelicaStandardLibrary/commit/adcbfc535c3b8ab0826eb21f59c268c40a473125, let's check if https://libraries.openmodelica.org/branches/master/Modelica_4.1.0-beta.om/Modelica_4.1.0-beta.om.html is improved when the test job runs again

**Comment by casella** [comment]
Yes, indeed all the 6 models mentioned in the first post now pass verification. Thanks!
